File: dairy_erp/dairy_erp/report/cattle_feed_advance_report/cattle_feed_advance_report.py
# Copyright (c) 2013, indictrans technologies and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import has_common

def execute(filters=None):
	columns, data = get_columns(), get_data(filters)
	return columns ,data 

def get_columns():

	columns = [
		_("Farmer Id") + ":Link/Farmer:150",
		_("Famer Name") + ":Data:200",
		_("Sales Invoice") + ":Link/Sales Invoice:150",
		_("Date Of Advance Taken") + ":Date:170", 
		_("Item Details") + ":Link/Item:150",
		_("Advance Amount") + ":Float:150",
		_("Cleared Amount") + ":Float:150",
		_("Pending/Balance") + ":Float:150",
		_("Remarks") + ":Data:200"
	]

	return columns

def get_data(filters):
	data = frappe.db.sql("""select
									si.customer,
									si.farmer,
									si.name,
									si.posting_date,
									SUBSTRING_INDEX(GROUP_CONCAT(si_item.item_name), ' ', 2),
									si.grand_total,
									si.grand_total - si.outstanding_amount,
									si.outstanding_amount,
									"",
									si.local_sale_type
							from
								`tabSales Invoice` si,
								`tabSales Invoice Item` si_item
							where
								si.local_sale = 1
								and si.name = si_item.parent
								and si.local_sale_type = 'Feed And Fodder Advance'
								and si.customer_or_farmer = "Farmer"
								and si_item.item_code not in ('COW Milk','BUFFALO Milk')
								and si.docstatus = 1 and si.company = '{0}'
								{1} GROUP BY si.name """.format(filters.get('vlcc'),get_conditions(filters)),filters,as_list=1,debug=0)

	if len(data):
		advance_data = []
		for row in data:
			if row[9] == "Feed And Fodder Advance" and len(row[2].split('-')) > 2:
				advance_data = get_feed_advance(row[2])
				if advance_data:
					row[5] = advance_data[0]
					row[6] = advance_data[1]
					row[7] = advance_data[2]
	return data


def get_feed_advance(invoice):
	if invoice:
		advance_name = frappe.get_value("Farmer Advance",{"feed_and_fodder_si":invoice},"name")
		if advance_name:
			advance_doc = frappe.get_doc("Farmer Advance",advance_name)
			return [advance_doc.advance_amount,advance_doc.advance_amount-advance_doc.outstanding_amount,advance_doc.outstanding_amount]

def get_conditions(filters):
	conditions = " and 1=1"
	if filters.get('farmer') and filters.get('start_date') and filters.get('end_date'):	
		conditions += " and si.posting_date between %(start_date)s and %(end_date)s and si.farmer = %(farmer)s"
	elif filters.get('start_date') and filters.get('end_date'):
		conditions += " and si.posting_date between %(start_date)s and %(end_date)s"
	return conditions

File: dairy_erp/dairy_erp/report/cattle_feed_advance_report/cattle_feed_advance_report.js
// Copyright (c) 2016, indictrans technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Cattle Feed Advance Report"] = {
	"filters": [
		{
			"fieldname":"vlcc",
			"label": __("VLCC"),
			"fieldtype": "Link",
			"options":"Company",
			"hidden": 1,
			"default":frappe.boot.user.first_name
		},
		{
			"fieldname":"farmer",
			"label": __("Farmer"),
			"fieldtype": "Link",
			"options":"Farmer",
			"get_query": function (query_report) {
				var vlcc_name = frappe.query_report_filters_by_name.vlcc.get_value();
				if(vlcc_name){
					return {
						"filters": {
							"vlcc_name": vlcc_name					
							}
						}
				}
			},
			"on_change": function(query_report) {
				console.log("insdie on_change")
				frappe.call({
					method: "frappe.client.get_value",
					args: {
						doctype: "Farmer",
						filters: {"name": frappe.query_report_filters_by_name.farmer.get_value()},
						fieldname: ["full_name"]
					},
					callback: function(r) {
						console.log("insidr e",r.message)
						if(!r.exc && r.message && !in_list(["Administrator", "Guest"], frappe.session.user)){
							if(has_common(frappe.user_roles, ["Vlcc Operator", "Vlcc Manager"])){
								frappe.query_report_filters_by_name.farmer_name.set_input(r.message.full_name);
							}
							query_report.trigger_refresh();
						}
					}
				})
			}
		},
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			"reqd":1,
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			"reqd":1,
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname":"vlcc_addr",
			"label": __("VLCC Address"),
			"fieldtype": "Data",
			"hidden": 1
		},
		{
			"fieldname":"farmer_name",
			"label": __("Farmer Name"),
			"fieldtype": "Data",
			"hidden": 0
		}	
	],
	onload: function(query_report) {
		frappe.call({
			method: "frappe.client.get_value",
			args: {
				doctype: "Village Level Collection Centre",
				filters: {"name": frappe.boot.user.first_name},
				fieldname: ["address_display"]
			},
			callback: function(r) {
				console.log("insidr e",r.message.address_display,r.message.address_display.split("<br>"))
				if(!r.exc && r.message && !in_list(["Administrator", "Guest"], frappe.session.user)){
					if(has_common(frappe.user_roles, ["Vlcc Operator", "Vlcc Manager"])){
						// $('body').find("[data-fieldname=vlcc]").val(r.message.company)
						//r.message.address_display = r.message.address_display.split("<br>")
						frappe.query_report_filters_by_name.vlcc_addr.set_input(r.message.address_display);
					}
					query_report.trigger_refresh();
				}
			}
		})
	}
}
